architecture/evm/evm_state_poller.go

```diff
@@ -27,6 +27,31 @@ const DefaultToleratedBlockHeadRollback = 1024
 
 var _ common.EvmStatePoller = &EvmStatePoller{}
 
+// helps prevent redundant concurrent polls for the same state
+type pollingMultiplexer struct {
+	result interface{}
+	err    error
+	done   chan struct{}
+	mu     sync.RWMutex
+	once   sync.Once
+}
+
+func newPollingMultiplexer() *pollingMultiplexer {
+	return &pollingMultiplexer{
+		done: make(chan struct{}),
+	}
+}
+
+func (m *pollingMultiplexer) Close(result interface{}, err error) {
```

Comment:
@aramalipoor I thought about re-using the multiplexer logic [here](https://github.com/erpc/erpc/blob/main/erpc/multiplexer.go), but the logic around multiplexing polls seemed different than network requests.

do you think having it as part of `state poller` makes sense? or should we think about having it more generic?

Reply:
They seem similar, only that "result" type is different. You could try a design with generics see if it's clean.

Having a generic Multiplexer (probably in `util` pkg) is good re testability and DRY.

Reply:
Even some of the usage parts could be abstracted away (`executeMultiplexedPoll` in here, or `handleMultiplexing` in Network)...

If it's not straightfoward, for now only having a generic-based Multiplexer object could be enough. Later we can refactor the usage in a separate PR.